### PR TITLE
release-22.2: roachtest: add mixed-version test for user login

### PR DESCRIPTION
**Pull request by ZhouXing19**
This is a follow up PR for #97183. We add tests for user login after upgrading from 22.1.14 to the current latest version.

informs #97178

Release note: None
Release justification: test only changes.

### CODE REVIEW

**Review by srosenberg — COMMENTED** [review]
Thanks for adding a regression roachtest! I only took a quick pass; will do a thorough one tomorrow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-:-NON9rSoEpn7OZmL4O1W:btemcvr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, and @ZhouXing19)

___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NON9TY-6lYKdSxgRl8T:-NON9TY-6lYKdSxgRl8U:bvik8iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/33294b285fd67c1c5e0d8c0b205a9a2dfb0492b8/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L13)):*
> ```Go
> )
> 
> func registerUserLoginMixedVersion(r registry.Registry) {
> ```

Could we add a comment here to note that this is a regression test for #97178? It'll make it easier to refactor/consolidate in the future.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NON9FGHB8FBkEHx9HTB:-NON9FGHB8FBkEHx9HTC:b-zaath1) ([raw file](https://github.com/cockroachdb/cockroach/blob/33294b285fd67c1c5e0d8c0b205a9a2dfb0492b8/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L29)):*
> ```Go
> 	const (
> 		mainVersion = ""
> 		preVersion  = "22.1.13"
> ```

Let's not hardcode the previous version to make this test future-proof. You can get the previous version relative to the current (build) version; please see `registerFixtures` and `makeVersionFixtureAndFatal`.

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NON9TY-6lYKdSxgRl8T:-NONOE8A27shj2dLBFAe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33294b285fd67c1c5e0d8c0b205a9a2dfb0492b8/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L13)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Could we add a comment here to note that this is a regression test for #97178? It'll make it easier to refactor/consolidate in the future.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NON9FGHB8FBkEHx9HTB:-NONOeYg8ikl1M9IVIvq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33294b285fd67c1c5e0d8c0b205a9a2dfb0492b8/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L29)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Let's not hardcode the previous version to make this test future-proof. You can get the previous version relative to the current (build) version; please see `registerFixtures` and `makeVersionFixtureAndFatal`.
</blockquote></details>

Done.

File: pkg/cmd/roachtest/tests/mixed_version_user_login.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		waitForUpgradeStep(c.All()),
+		checkCurrentVersionStep,
+		// Create a new user and set cluster.preserve_downgrade_option.
+		createUserStep,
+		setPreserveDowngradeOptionStep,
+		// Upgrade all nodes to the latest version.
+		binaryUpgradeStep(c.All(), mainVersion),
+		checkCurrentVersionStep,
+		// Connect to the servers as the new user.
+		connAsNewUserStep,
```
Comment (renatolabs) [inline pkg/cmd/roachtest/tests/mixed_version_user_login.go:98]: I know this is sufficient to reproduce the regression, but it wouldn't hurt to let the upgrade finalize and check that you can log in again.

File: pkg/cmd/roachtest/tests/mixed_version_user_login.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 The Cockroach Authors.
```
Comment (renatolabs) [inline pkg/cmd/roachtest/tests/mixed_version_user_login.go:1]: It probably doesn't matter, but should we reference 2023 here?

**Review by renatolabs — COMMENTED** [review]
Thank you!

Are you planning to forward-port this test? We have been doing some work recently in the upgrade testing space, and I have some ideas about how you could write this test on master (taking #92431 into account). Let me know!

**Review by ZhouXing19 — COMMENTED** [review]
Yeah I do plan to fore-port this. Thanks for the pointer to the new infra! Will refer to it when work on the master branch.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-:-NOQbAua4P0IEgHZVo50:b-jf764w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-1:-NOQa60A0IVfda0Cva5b:balrtrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L1)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

It probably doesn't matter, but should we reference 2023 here?
</blockquote></details>

Oh haha I stayed in the past ... Fixed 
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-47:-NOQaA3ACL0_iBUFb-9F:b-lr3wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L47)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I guess this is not technically wrong, but this is not a `testing.T` test, so using `require` can lead to more confusing error messages. I'd recommend replacing use of `require` here with calls to `t.Fatal`.
</blockquote></details>

TIL! Done.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-54:-NOQaCvi-GM6oS5kxV5T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

You could use `u.conn()` directly here (and everywhere other than `createUserStep`) and let the framework deal with closing connections at the end of the test.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-88:-NOQbSmU-TbDpdPBYRg5:ba3yf8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L88)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Should we start with the fixtures for this version? See `uploadAndStartFromCheckpointFixture`.
</blockquote></details>

I tried adding 
```
		uploadAndStartFromCheckpointFixture(c.All(), prevStableVersionStr),
		waitForUpgradeStep(c.All()),
```
of the test flow, but it hanged with 
```
19:39:32 cockroach.go:667: jane-1676576236-01-n1cpu4: COCKROACH_DEV_LICENSE unset: enterprise features will be unavailable
```
for couple of minutes. 
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-93:-NOQaXnIEpRW3oHuyeeg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L93)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

You can use the existing `preventAutoUpgradeStep` for this purpose.

https://github.com/cockroachdb/cockroach/blob/release-22.2/pkg/cmd/roachtest/tests/versionupgrade.go#L471-L479
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-98:-NOQaTBD27Iqc1M6vJDA:b-wrwfa0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L98)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I know this is sufficient to reproduce the regression, but it wouldn't hurt to let the upgrade finalize and check that you can log in again.
</blockquote></details>

Yes, without this `waitForUpgradeStep` it's sufficient to repro the regression.
I tried adding `waitForUpgradeStep(c.All()), connAsNewUserStep` here, but only got this (timeout) error
```
18:18:27 versionupgrade.go:503: 22.2: waiting for cluster to auto-upgrade
18:23:28 test_impl.go:342: test failure #1: (test_impl.go:286).Fatal: 1: expected version 22.2, got 22.1
```
with running it couple of times. I also tried running `select version()` at this step against the node 1, and got 22.4.

**Review by renatolabs — APPROVED** [review]
Great! Let me know if you have questions or any issues in that process.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-:-NOQxQ_y2eTpAy8IFXAz:b-mbxi4u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, @srosenberg, and @ZhouXing19)

___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-88:-NOQwyai3IxvIUT8g7WW:b7jgef7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L88)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

I tried adding 
```
		uploadAndStartFromCheckpointFixture(c.All(), prevStableVersionStr),
		waitForUpgradeStep(c.All()),
```
of the test flow, but it hanged with 
```
19:39:32 cockroach.go:667: jane-1676576236-01-n1cpu4: COCKROACH_DEV_LICENSE unset: enterprise features will be unavailable
```
for couple of minutes. 
</blockquote></details>

Did you replace `binaryUpgradeStep` with `uploadAndStartFromCheckpointFixture`? Weird, haven't seen that before.
___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-98:-NOQxAOB8U2xVmcQQk06:b1id5wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L98)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Yes, without this `waitForUpgradeStep` it's sufficient to repro the regression.
I tried adding `waitForUpgradeStep(c.All()), connAsNewUserStep` here, but only got this (timeout) error
```
18:18:27 versionupgrade.go:503: 22.2: waiting for cluster to auto-upgrade
18:23:28 test_impl.go:342: test failure #1: (test_impl.go:286).Fatal: 1: expected version 22.2, got 22.1
```
with running it couple of times. I also tried running `select version()` at this step against the node 1, and got 22.4. 
</blockquote></details>

`waitForUpgradeStep` will just wait for the cluster version to reach a certain version, you need to reset the "preserve downgrade" option before you do that. You can call `allowAutoUpgradeStep`.

**Comment by rafiss** [comment]
i was able to write a reproducing test with the logic test framework like this:
```
# LogicTest: cockroach-go-testserver-22.1-22.2

query T
SELECT crdb_internal.node_executable_version()
----
22.1

upgrade 0

upgrade 1

user testuser nodeidx=0

query T
SELECT crdb_internal.node_executable_version()
----
22.2
```

however, this framework is only available on `master`, so this PR lgtm!

**Review by ZhouXing19 — COMMENTED** [review]
Nice, I'll try using the logictest framework on master as well!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-:-NOR0HqC4XOlMKDbJ6nr:b-evt9m8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-88:-NOR03x3CFrw2aaT9lqv:bx5uey2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L88)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Did you replace `binaryUpgradeStep` with `uploadAndStartFromCheckpointFixture`? Weird, haven't seen that before.
</blockquote></details>

Yes, the stack starts with `uploadAndStartFromCheckpointFixture`, and it's stuck at this step.
```
20:56:14 versionupgrade.go:199: test status: versionUpgradeTest: starting step 1
20:56:15 cluster.go:677: test status: uploading file
20:56:15 cluster_synced.go:1600: jane-1676580856-01-n1cpu4: putting (dist) pkg/cmd/roachtest/fixtures/1/checkpoint-v22.1.tgz {store-dir}/fixture.tgz on nodes [1]
20:56:16 cluster_synced.go:1798: 
20:56:16 cluster_synced.go:1801:    1: done
20:56:16 cluster.go:677: test status: 
20:56:19 cluster.go:677: test status: staging binary
20:56:19 staging.go:296: Resolved release url for cockroach version v22.1.13: https://storage.googleapis.com/cockroach-release-artifacts-prod/cockroach-v22.1.13.linux-amd64.tgz
20:56:23 cluster.go:677: test status: 
20:56:23 cluster.go:677: test status: starting nodes :1
20:56:23 cockroach.go:158: jane-1676580856-01-n1cpu4: starting nodes
20:56:26 cockroach.go:207: jane-1676580856-01-n1cpu4: setting cluster settings
20:56:26 cockroach.go:667: jane-1676580856-01-n1cpu4: COCKROACH_DEV_LICENSE unset: enterprise features will be unavailable
<stuck>
```

**Comment by ZhouXing19** [comment]
*[`pkg/cmd/roachtest/tests/mixed_version_user_login.go` line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97220#-NOMC0aZJeIK6IAEGpGH-r2-98:-NOR1i732znBNQC8229N:b-jgitj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e53d18bce66ef836f315a3fe62231feac5bc8c10/pkg/cmd/roachtest/tests/mixed_version_user_login.go#L98)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

`waitForUpgradeStep` will just wait for the cluster version to reach a certain version, you need to reset the "preserve downgrade" option before you do that. You can call `allowAutoUpgradeStep`.
</blockquote></details>

Yup with `allowAutoUpgradeStep` it worked. Updated.

**Review by srosenberg — APPROVED** [review]
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97220)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @rafiss, and @renatolabs)

**Comment by ZhouXing19** [comment]
TFTR!

**Pull request merged**